main/app.c

```diff
@@ -2306,16 +2623,194 @@ int ast_app_group_update(struct ast_channel *old, struct ast_channel *new)
 	return 0;
 }
 
+int ast_app_group_add_meta_ifneeded(const char *group, const char *category)
+{
+	struct ast_group_meta *gmi = NULL; /*!< Group metadatas */
+
+	AST_RWLIST_WRLOCK(&groups_meta);
+	AST_RWLIST_TRAVERSE_SAFE_BEGIN(&groups_meta, gmi, group_meta_list) {
+		if (!strcasecmp(gmi->group, group) && !strcasecmp(gmi->category, category)) {
+			break; /* We only have one list item per group@category */
+		}
+	}
+	AST_RWLIST_TRAVERSE_SAFE_END;
+
+	if (!gmi) {
+		if (!(gmi = ast_calloc(1, sizeof(struct ast_group_meta)))) {
+		    AST_RWLIST_UNLOCK(&groups_meta);
+		    return -1;
+		}
+
+		ast_copy_string(gmi->group, group, MAX_GROUP_LEN);
+
+		if (!ast_strlen_zero(category)) {
+			ast_copy_string(gmi->category, category, MAX_CATEGORY_LEN);
+		}
+
+		AST_RWLIST_INSERT_TAIL(&groups_meta, gmi, group_meta_list);
+
+		manager_event(EVENT_FLAG_DIALPLAN, "GroupCreate",
+			"Category: %s\r\n"
+			"Group: %s\r\n",
+			category, group);
+	}
+
+	gmi->num_channels++;
+
+	AST_RWLIST_UNLOCK(&groups_meta);
+
+	return 1;
+}
+
+/* Remove all channels from the given group */
+int ast_app_group_remove_all_channels(const char *group, const char *category)
+{
+	struct ast_group_info *gi = NULL;
+
+	int group_meta_found = 0;
+
+	if (!group) {
+		ast_log(LOG_WARNING, "Group Remove: At the minimum, group must be non-NULL\n");
+		return 0;
+	}
+
+	if (!category) {
+		category = "";
+	}
+
+	/* Traverse group@category channel assignments */
+	AST_RWLIST_WRLOCK(&groups);
+	AST_RWLIST_TRAVERSE_SAFE_BEGIN(&groups, gi, group_list) {
+		if (strcasecmp(gi->group, group) || strcasecmp(gi->category, category)) {
+			/* Need to match group@category exactly */
+			continue;
+		}
+
+		manager_event(EVENT_FLAG_DIALPLAN, "GroupChannelRemove",
+			"Channel: %s\r\n"
+			"Category: %s\r\n"
+			"Group: %s\r\n"
+			"Uniqueid: %s\r\n",
+			ast_channel_name(gi->chan),
+			gi->category, gi->group,
+			ast_channel_uniqueid(gi->chan));
+
+		ast_app_group_remove_meta_ifneeded(gi->group, gi->category);
+
+		AST_RWLIST_REMOVE_CURRENT(group_list);
+		ast_free(gi);
+
+		group_meta_found++;
+	}
+	AST_RWLIST_TRAVERSE_SAFE_END;
+	AST_RWLIST_UNLOCK(&groups);
+
+	return group_meta_found;
+}
+
+int ast_app_group_rename(const char *old_group, const char *old_category, const char *new_group, const char *new_category)
+{
+	struct ast_group_info *gi = NULL;
+	struct ast_group_meta *gmi = NULL; /*!< Group metadatas	*/
+
+	int groups_found = 0;
+
+	if (!old_group) {
+		ast_log(LOG_WARNING, "Old Group name is required for group rename\n");
+		return -1;
+	}
+
+	if (!new_group) {
+		ast_log(LOG_WARNING, "New Group name is required for group rename\n");
+		return -1;
+	}
+
+	if (!old_category) {
+		old_category = "";
+	}
+
+	if (!new_category) {
+		new_category = "";
+	}
+
```

Comment:
I think you could remove the empty lines between the preceding blocks and eliminate this empty line

Reply:
Empty line removed, but.. I think this is just a syle-preference at this point.

Is there anything in the coding style guidelines that requires blocks to be smushed together?  My eyes prefer them split apart!